---
title: "isoform switch"
author: "Jared Slosberg"
date: "6/22/2023"
output: html_document
---

Script needs to be run with R = 4.2. Use /opt/R/4.2.3/bin/R with libraries @ "/home/jared/R/x86_64-pc-linux-gnu-library/4.2".
Singularity container with isoformAnalysis.yml
Saturn = 1.7.3
IsoformSwitchAnalyzer = 2.1.2

```{r setup, include=FALSE}
.libPaths("/home/jared/R/x86_64-pc-linux-gnu-library/4.2")

knitr::opts_chunk$set(echo = TRUE)

```

```{r}
#library(tidyverse)
library(here)
library(IsoformSwitchAnalyzeR)
library(dplyr)
```
From the isoform quantification IsoformSwitchAnalyzeR performs five high-level tasks:

- Statistical identification of isoform switches.

- Integration of a wide range of (predicted) annotations for the isoforms involved in the identified switches (e.g. protein domains).

- Identification of which isoforms have a predicited functional consequnce (e.g. loss/gain of protein domain).

- Visualization of predicted consequences of the isoform switches for individual genes

- Analysis of genome wide patterns in both switch consequences and alternative splicing.
```{r}
# Load example data and pre-filter
data("exampleSwitchList")
exampleSwitchList <- preFilter(exampleSwitchList) # preFilter to remove lowly expressed features
#> The filtering removed 55 ( 21.24% of ) transcripts. There is now 204 isoforms left

# Perform test
exampleSwitchListAnalyzed <- isoformSwitchTestDEXSeq(
    switchAnalyzeRlist = exampleSwitchList,
    reduceToSwitchingGenes=TRUE
)
#> Step 1 of 2: Testing each pairwise comparisons with DEXSeq (this might be a bit slow)...
#>     Estimated run time is: 1 min
#> Step 2 of 2: Integrating result into switchAnalyzeRlist...
#>     Isoform switch analysis was performed for 60 gene comparisons (100%).
#> Total runtime: 0.14 min
#> Done

# Summarize switching features
extractSwitchSummary(exampleSwitchListAnalyzed)
#>            Comparison nrIsoforms nrSwitches nrGenes
#> 1 hESC vs Fibroblasts         82         58      38
```



```{r}
metadata <- read.csv(here("metadata_merged.csv")) %>% filter(sample_name != "17mo-M-1")
design <- metadata %>% transmute(sampleID = sample_name, condition = age) %>% mutate(
  condition = paste0("age_",condition)
)

kallistoQuant <- importIsoformExpression(
    parentDir = here("kallisto_out_merged/"),
    addIsofomIdAsColumn = TRUE
)

aSwitchList <- importRdata(
    isoformCountMatrix   = kallistoQuant$counts,
    isoformRepExpression = kallistoQuant$abundance,
    designMatrix         = design, 
    isoformExonAnnoation = here("isoform_level_analysis/gencode.vM27.annotation.gtf"),
    showProgress = T
)
```

```{r}

#lowered geneExpressionCutoff to include BDNF
aSwitchList_filt <- preFilter(aSwitchList, geneExpressionCutoff = 0.28, isoformExpressionCutoff = 0, removeSingleIsoformGenes = T)

SwitchRes <- isoformSwitchTestDEXSeq(aSwitchList_filt)

#saveRDS(SwitchRes, here("isoform_level_analysis/isoformSwitchAnalyzer/DEXSeq/DEXSeq_res.rds"))
```
```{r}
condition_list <- list("17mo_vs_P30" = c("cond1" = "age_17mo", "cond2" = "age_P30"),
                       "17mo_vs_6mo" = c("cond1" = "age_17mo", "cond2" = "age_6mo"),
                       "6mo_vs_P30" = c("cond1" = "age_6mo", "cond2" = "age_P30"))

pl_list <- purrr::map(names(condition_list), function(cond){
  print(cond)
  print(condition_list[[cond]][1])
  
  pdf(paste0(here("isoform_level_analysis/isoformSwitchAnalyzer/satuRn/BDNF_isoforms_"), cond, ".pdf"), onefile = F)

  
  switchPlot(
    SwitchRes,
    gene='Bdnf',
    IFcutoff = 0,
    condition1 = condition_list[[cond]][1],
    condition2 = condition_list[[cond]][2],
    rescaleTranscripts = F# making text sightly larger for vignette
  )
  
 dev.off()


  
})

ages <- c("age_17mo","age_6mo")
iso_fraction_df <- purrr::map_df(ages, function(age){
iso_fractions <- SwitchRes$isoformFeatures %>% filter(gene_id == "Bdnf", condition_1 == age) %>% dplyr::select(gene_id, isoform_id, IF1) %>% unique %>% mutate(age = age, IF = IF1) %>% dplyr::select(gene_id, age, isoform_id, IF)

})

iso_fractions <- SwitchRes$isoformFeatures %>% filter(gene_id == "Bdnf", condition_2 == "age_P30") %>% dplyr::select(gene_id, isoform_id, IF2) %>% unique %>% mutate(age = "age_P30", IF = IF2) %>% dplyr::select(gene_id, age,  isoform_id, IF)

iso_fraction_df <- rbind(iso_fraction_df, iso_fractions) %>% 
  mutate(pct = paste0(round(IF * 100, 1), " %")) %>% 
  mutate(pct = str_replace_all(pct, "0 %", ""))

pl <- ggplot(iso_fraction_df, aes(x = age, y = IF, fill = isoform_id)) + geom_col(position = "fill", color = "black") + ggtitle("Isoform Expression of BDNF") + ylab("Isoform fraction") + geom_text(aes(label = pct), position = "stack", vjust = 1.1, size= 3)
pl

pdf(here("isoform_level_analysis/isoformSwitchAnalyzer/satuRn/BDNF_isoforms_all_ages.pdf"))
pl
dev.off()
```

```{r}
pl_list <- purrr::map(names(condition_list), function(cond){
  print(cond)
  print(condition_list[[cond]][1])
  
  pdf(paste0(here("isoform_level_analysis/isoformSwitchAnalyzer/satuRn/TrkB_isoforms_"), cond, ".pdf"), onefile = F)

  
  switchPlot(
    SwitchRes,
    gene='Ntrk2',
    IFcutoff = 0,
    condition1 = condition_list[[cond]][1],
    condition2 = condition_list[[cond]][2],
    rescaleTranscripts = F
  )
  
 dev.off()


  
})

#todo: easier to grab IF values by condition from SwitchRes$isoformRepIF
ages <- c("age_17mo","age_6mo")
iso_fraction_df <- purrr::map_df(ages, function(age){
iso_fractions <- SwitchRes$isoformFeatures %>% filter(gene_id == "Ntrk2", condition_1 == age) %>% dplyr::select(gene_id, isoform_id, IF1) %>% unique %>% mutate(age = age, IF = IF1) %>% dplyr::select(gene_id, age, isoform_id, IF)

})

iso_fractions <- SwitchRes$isoformFeatures %>% filter(gene_id == "Ntrk2", condition_2 == "age_P30") %>% dplyr::select(gene_id, isoform_id, IF2) %>% unique %>% mutate(age = "age_P30", IF = IF2) %>% dplyr::select(gene_id, age,  isoform_id, IF)

iso_fraction_df <- rbind(iso_fraction_df, iso_fractions) %>% 
  mutate(pct = paste0(round(IF * 100, 1), " %")) %>% 
  mutate(pct = ifelse( IF > .03, pct, ""))

pl <- ggplot(iso_fraction_df, aes(x = age, y = IF, fill = isoform_id)) + geom_col(position = "fill", color = "black") + ggtitle("Isoform Expression of TrkB") + ylab("Isoform fraction") + geom_text(aes(label = pct), position = "stack", vjust = 1.1, size= 3)
pl

pdf(here("isoform_level_analysis/isoformSwitchAnalyzer/satuRn/TrkB_isoforms_all_ages.pdf"))
pl
dev.off()
```

```{r}
  
  switchPlot(
    DEXSeq_res,
    gene='Tbc1d10a',
    IFcutoff = 0,
    condition1 = "age_17mo",
    condition2 = "age_P30",
    rescaleTranscripts = F
  )
```

```{r}
BDNF_switchlist <- subsetSwitchAnalyzeRlist(
    switchAnalyzeRlist = aSwitchList_filt,
    subset =  aSwitchList_filt$isoformFeatures$gene_id == "Bdnf"
)

#todo: easier to grab IF values by condition from SwitchRes$isoformRepIF
ages <- c("age_17mo","age_6mo")
iso_fraction_df <- purrr::map_df(ages, function(age){
iso_fractions <- BDNF_switchlist$isoformFeatures %>% filter(gene_id == "Bdnf", condition_1 == age) %>% dplyr::select(gene_id, isoform_id, IF1) %>% unique %>% mutate(age = age, IF = IF1) %>% dplyr::select(gene_id, age, isoform_id, IF)

})

iso_fractions <- BDNF_switchlist$isoformFeatures %>% filter(gene_id == "Bdnf", condition_2 == "age_P30") %>% dplyr::select(gene_id, isoform_id, IF2) %>% unique %>% mutate(age = "age_P30", IF = IF2) %>% dplyr::select(gene_id, age,  isoform_id, IF)

iso_fraction_df <- rbind(iso_fraction_df, iso_fractions) %>% 
  mutate(pct = paste0(round(IF * 100, 1), " %")) %>% 
  mutate(pct = ifelse( IF > .03, pct, ""))


```
Mods for plotting
```{r}
bdnf_desc <- read.csv(here("isoform_level_analysis/data/bdnf_isoform_description.csv"))

iso_fractions_df <- iso_fractions_df %>% mutate(
    age = case_when(
      age == "age_17mo" ~ "17mo",
      age == "age_6mo" ~ "6mo",
      age == "age_P30" ~ "1mo")) %>%
  mutate(age = factor(age, levels = c("1mo","6mo","17mo")))

iso_fractions_df <- iso_fractions_df %>%
  tidyr::separate(isoform_id, into = "transcript_id", sep = "\\.", remove = F) %>% 
  left_join(bdnf_desc, by = "transcript_id") %>% 
  mutate(BDNF_isoform = paste0(isoform, " (",exon_id,")"))

iso_fractions_df <- left_join(iso_fractions_df, bdnf_desc)

pl <- ggplot(iso_fractions_df, aes(x = age, y = IF, fill = BDNF_isoform)) +
  geom_col(position = "fill", color = "black") +
  ggtitle("Isoform Expression of BDNF") +
  ylab("Isoform fraction") +
  geom_text(aes(label = pct), position = position_stack(vjust = .5), size= 3) +
  theme_bw() +
  xlab("Age") +
  scale_fill_brewer(palette = "Set3")


write.csv(iso_fraction_df, here("isoform_level_analysis/data/bdnf_isoform_fractions.csv"))

pdf(here("isoform_level_analysis/isoformSwitchAnalyzer/satuRn/bdnf_isoforms_all_ages.pdf"))
pl
dev.off()
```

