---
title: "Identify groups of DE genes with similar dynamics"
output: html_document
---


```{r, include= F}
library(tidyverse)
library(glue)
library(DESeq2)
library(here)
library(tximport)
library(RColorBrewer)
library(vsn)
library(ComplexHeatmap)
library(biomaRt)
```

Read in experimental metadata
```{r}
data_dir <- here("kallisto_out_merged/")
metadata_path <- here("metadata_merged_filtered.csv")
list.files(data_dir)


exp_metadata <- read.csv(metadata_path, header = T)

# named_runs <- exp_metadata %>% deframe
```



Differential expression on relevant conditions
```{r}
dds <- readRDS(here("data/bulkrna_dds.rds"))

test_names <- paste0("age_",c("6mo","17mo"),"_vs_P30")

res_list <- map(test_names, function(name){
  tbl <- read.csv(paste0(here("results/diff_exp/"), name,".csv"), row.names = 1)
})

clip_nlog10p <- 10
sigthresh_nlog10p <- -log10(.05)
res_df <- map_df(test_names, function(name){
  tbl <- read.csv(paste0(here("results/diff_exp/"), name,".csv"), row.names = 1)
  tbl <- tbl %>% mutate(condition = name,
                        nlog10p = ifelse(-log10(padj) > clip_nlog10p, clip_nlog10p, -log10(padj)),
                        nlog10p_clipped = ifelse(-log10(padj) > clip_nlog10p, T, F ))
  return(tbl)
})


```



Get biomart annotations for gene short names
```{r}
library(biomaRt)
#get gene names, make sure to use the right ensembl archive for genome used for alignment
#listMarts() 
ensembl <- useMart("ensembl")
datasets <- listDatasets(ensembl)
ensembl <- useDataset("mmusculus_gene_ensembl",mart=ensembl)

gene_id_trimmed <- rownames(dds)
gene_description <- getBM(attributes = c("ensembl_gene_id_version","external_gene_name","description","chromosome_name"),
                   filters = "ensembl_gene_id", values =gene_id_trimmed,  mart = ensembl)

geneid2name <- gene_description[,c("ensembl_gene_id_version", "external_gene_name")] %>% 
  mutate(ensembl_gene_id_trimmed = str_split_fixed(ensembl_gene_id_version, "\\.", 2) %>% .[,1])

gene_ids <- geneid2name %>% mutate(ensemblgene_id = ensembl_gene_id_version)
```

```{r}
pcut <- 0.05
#all genes that are DE in either condition
diff_genes <-res_df %>% filter(padj < pcut) %>% pull(ensemblgene_id) %>% unique()

res_df_filtered <- res_df %>% filter(ensemblgene_id %in% diff_genes)
res_df_filtered_wide <- res_df_filtered  %>%
  dplyr::select(c(ensemblgene_id, padj, nlog10p, condition, log2FoldChange)) %>%
  pivot_wider(names_from = condition, values_from = c(padj, log2FoldChange, nlog10p)) %>% 
  mutate(both_sig = padj_age_6mo_vs_P30 < pcut & padj_age_17mo_vs_P30<pcut,
         doe_17mo_vs_P30 = sign(log2FoldChange_age_17mo_vs_P30),
         doe_6mo_vs_P30 = sign(log2FoldChange_age_6mo_vs_P30),
         direction_consistent = (doe_17mo_vs_P30 == doe_6mo_vs_P30),
         direction = ifelse(direction_consistent, sign(log2FoldChange_age_6mo_vs_P30), paste(doe_6mo_vs_P30,doe_17mo_vs_P30, sep = ","))) 

#for genes significant for both timepoints, what are the direction of effects?
res_df_filtered_wide %>% filter(both_sig == T) %>% pull(direction) %>% table

#for genes significant for at least one timepoint, what are the direction of effects (over both timepoints)?
res_df_filtered_wide %>% pull(direction) %>% table

genes_down <- res_df_filtered_wide %>% filter(both_sig == T, direction == -1) %>% pull(ensemblgene_id) 
genes_up <- res_df_filtered_wide %>% filter(both_sig == T, direction == 1) %>% pull(ensemblgene_id) 

```


```{r}

cluster_df <- read.csv(here("results/dynamic_gene_modules.csv"))

log2fc_long <- res_df_filtered %>%
  full_join(cluster_df) %>%
  mutate(condition = 
           case_when(condition == "age_6mo_vs_P30" ~ "6mo", condition == "age_17mo_vs_P30" ~ "17mo")) %>%
  mutate(transparency = 2*sqrt(1/n)) %>%
  mutate(transparency_corrected = pmin(1, transparency)) %>% 
  dplyr::select(-c(baseMean, lfcSE, stat, pvalue, padj, nlog10p, nlog10p_clipped))

log2fc_mat <- res_df_filtered_wide %>% dplyr::select(ensemblgene_id, log2FoldChange_age_6mo_vs_P30, log2FoldChange_age_17mo_vs_P30) %>% mutate(log2FoldChange_age_P30_vs_P30 = 0) %>% column_to_rownames(var = "ensemblgene_id") %>% as.matrix()
#scale but don't center. Positive values should remain positive, negative remain negative. L2 norm
scaled_log2fc_mat <- t(scale(t(log2fc_mat), scale = T, center = F))


```

```{r}
p30_df <- log2fc_long %>% dplyr::select(external_gene_name, ensemblgene_id, sph_cluster, n, cluster_title, transparency, transparency_corrected) %>% unique %>% mutate("log2FoldChange" = 0, condition = "P30")

log2fc_long <- rbind(log2fc_long, p30_df) %>% mutate(condition = factor(condition, levels = c("P30","6mo","17mo"))) 
dummy_df <- log2fc_long %>% dplyr::select(-cluster_title)


scaled_log2fc_long <- scaled_log2fc_mat %>% as.data.frame() %>%
  rownames_to_column(var = "ensemblgene_id") %>%
  pivot_longer(!ensemblgene_id, values_to = "scaled_log2fc", names_to = "comp") %>%
  mutate(condition = factor(case_when(
    comp == "log2FoldChange_age_P30_vs_P30" ~ "P30",
    comp == "log2FoldChange_age_6mo_vs_P30" ~ "6mo",
    comp == "log2FoldChange_age_17mo_vs_P30" ~ "17mo",
    ), levels = c("P30","6mo","17mo"))) %>%
  full_join(cluster_df) %>%
  mutate(transparency = sqrt(1/n)) %>%
  mutate(transparency_corrected = pmin(1, transparency))


pdf(here("plots/figures/de_gene_module_dynamics.pdf"), width = 9)
ggplot(scaled_log2fc_long, aes( x = condition, y = scaled_log2fc, group = ensemblgene_id, alpha = transparency_corrected)) +
  geom_line(color = "black") +
  facet_wrap(~cluster_title) +
  geom_hline(yintercept = 0, linetype = "dashed", color = "grey50")+
  ylab("scaled log2FC estimates relative to P30") +
  ggtitle("clustered on cosine distance of LFC")
dev.off()

```


```{r}
res_df_filtered_wide_clusters <- res_df_filtered_wide %>% left_join(cluster_df)

p4 <- ggplot(res_df_filtered_wide_clusters,aes(x = log2FoldChange_age_6mo_vs_P30, y=log2FoldChange_age_17mo_vs_P30, color = cluster_title)) + geom_point(size = I(0.01))+ xlab("6mo vs P30 log2FC") + ylab("17mo vs P30 log2FC") + ggtitle("Subset of genes for which at least one condition is signficantly DE")+ guides(colour = guide_legend(override.aes = list(size=5)))

p5 <- ggplot(res_df_filtered_wide_clusters,aes(x = log2FoldChange_age_6mo_vs_P30, y=log2FoldChange_age_17mo_vs_P30, color = cluster_title)) + geom_point(size = I(0.01))+ xlab("6mo vs P30 log2FC") + ylab("17mo vs P30 log2FC") + ggtitle("Subset of genes for which at least one condition is signficantly DE")+ guides(colour = guide_legend(override.aes = list(size=5))) + xlim(c(-3,3)) + ylim(c(-3,3)) 


p2

p4
p5

table(res_df_filtered_wide_clusters[,c("direction","cluster_title")])



```
