---
title: "aged timecourse bulkRNA-seq "
output: html_document
---


```{r, include= F}
library(tidyverse)
library(glue)
library(DESeq2)
library(here)
library(tximport)
library(RColorBrewer)
library(vsn)
library(ComplexHeatmap)
library(biomaRt)
```

Read in experimental metadata
```{r}
data_dir <- here("kallisto_out_merged/")
metadata_path <- here("metadata_merged.csv")
list.files(data_dir)


exp_metadata <- read.csv(metadata_path, header = T)

abundance_filepath <- paste0(data_dir, exp_metadata[,"sample_name"], "/abundance.tsv")
names(abundance_filepath) <- exp_metadata[, "sample_name"]
file.exists(abundance_filepath)

# named_runs <- exp_metadata %>% deframe
```

Read in gene abundance matrices for each run
```{r}
txdb <- GenomicFeatures::makeTxDbFromGFF("/reference/genomes/mouse/gencode/vM27/gencode.vM27.annotation.gff3.gz")
k <- keys(txdb, keytype = "TXNAME" )
tx2gene <- select(txdb, k, "GENEID", "TXNAME")

#There is an issue with how kallisto writes its .h5 files (https://support.bioconductor.org/p/130419/)
#So use tsv, and dropInfReps=T to avoid anything related to the .h5 file
txi <- tximport(abundance_filepath, type = "kallisto", txOut = F, tx2gene = tx2gene, ignoreAfterBar = T, dropInfReps = T, countsFromAbundance = "scaledTPM")

#write.csv(txi$counts, here("data/abundance_to_counts_matrix.csv"), quote = F)
```

Check transformations from abundance to counts
```{r}
#output from kallisto quantification
txi$abundance %>% head

txi$counts %>% head

#qplot(log2(txi$abundance), log2(txi$counts), color = txi$length) + scale_color_gradient(name = "tx length", trans = "log2",breaks = c(500, 10000, 30000, 60000, 120000), labels = c(500, 10000, 30000, 60000, 120000))
```




Get biomart annotations for gene short names
```{r}
library(biomaRt)
#get gene names, make sure to use the right ensembl archive for genome used for alignment
#listMarts() 
ensembl <- useMart("ensembl")
datasets <- listDatasets(ensembl)
ensembl <- useDataset("mmusculus_gene_ensembl",mart=ensembl)

gene_id_trimmed <- stringr::str_split_fixed(unique(tx2gene[,"GENEID"]),"\\.",2)[,1]
gene_description <- getBM(attributes = c("ensembl_gene_id_version","external_gene_name","description","chromosome_name"),
                   filters = "ensembl_gene_id", values =gene_id_trimmed,  mart = ensembl)

geneid2name <- gene_description[,c("ensembl_gene_id_version", "external_gene_name")] %>% 
  mutate(ensembl_gene_id_trimmed = str_split_fixed(ensembl_gene_id_version, "\\.", 2) %>% .[,1])

```

Following DESEQ tutorial @ [here](http://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)

Prep data frame with sample <<-->> group info
```{r}
# #experimental design
# cond <- colnames(txi$abundance)
# 
# coldata <- data.frame("condition" = colnames(total_counts)) %>%
#   mutate(condition = case_when(
#     grepl("HF_", condition) ~ "HF",
#     grepl("NF_", condition) ~ "NF",
#     grepl("NF-sugar_", condition) ~ "NF-sugar",
#     grepl("HF-sugar_", condition) ~ "HF-sugar") %>%
#       as.factor()
#   ) 

coldata <- exp_metadata %>% mutate(tech.rep.collapsed = paste(age, sex, rep, sep = "-"))

assertthat::assert_that(all(coldata$sample_name == colnames(txi$abundance)))

coldata 
#Batches or which tissues are from the same animal are unknown



```
Create DeSeq object with counts and exp design. Don't include batch at first
```{r}

dds <- DESeqDataSetFromTximport(txi,
                              colData = coldata,
                              design = ~ age)

dds$age <- factor(dds$age, levels = c("P30","6mo","17mo"), ordered = F)


```

### Now read in the processing with unmerged tech reps

```{r}
data_dir <- here("kallisto_out/")
metadata_path <- here("metadata.csv")
list.files(data_dir)


exp_metadata_all <- read.csv(metadata_path, header = T)

abundance_filepath <- paste0(data_dir, exp_metadata_all[,"sample_name"], "/abundance.tsv")
names(abundance_filepath) <- exp_metadata_all[, "sample_name"]
file.exists(abundance_filepath)

# named_runs <- exp_metadata %>% deframe
```

Read in gene abundance matrices for each run
```{r}
#There is an issue with how kallisto writes its .h5 files (https://support.bioconductor.org/p/130419/)
#So use tsv, and dropInfReps=T to avoid anything related to the .h5 file
txi_all <- tximport(abundance_filepath, type = "kallisto", txOut = F, tx2gene = tx2gene, ignoreAfterBar = T, dropInfReps = T, countsFromAbundance = "scaledTPM")

#write.csv(txi$counts, here("data/abundance_to_counts_matrix.csv"), quote = F)
```

Prep data frame with sample <<-->> group info
```{r}
# #experimental design
# cond <- colnames(txi$abundance)
# 
# coldata <- data.frame("condition" = colnames(total_counts)) %>%
#   mutate(condition = case_when(
#     grepl("HF_", condition) ~ "HF",
#     grepl("NF_", condition) ~ "NF",
#     grepl("NF-sugar_", condition) ~ "NF-sugar",
#     grepl("HF-sugar_", condition) ~ "HF-sugar") %>%
#       as.factor()
#   ) 

coldata <- exp_metadata_all  %>% 
  mutate(tech.rep.collapsed = paste(age, sex, rep, sep = "-")) 

assertthat::assert_that(all(coldata$sample_name == colnames(txi_all$abundance)))

coldata 
#Batches or which tissues are from the same animal are unknown



```
Create DeSeq object with counts and exp design. Don't include batch at first
```{r}

dds_all <- DESeqDataSetFromTximport(txi_all,
                              colData = coldata,
                              design = ~ age)

dds_all$age <- factor(dds_all$age, levels = c("P30","6mo","17mo"), ordered = F)

dds_all <- collapseReplicates(dds_all, groupby = coldata$tech.rep.collapsed, run = rownames(coldata))

```

```{r}
counts_merged_before <- counts(dds) %>% as.data.frame() %>% tibble::rownames_to_column(var = "gene_id") %>%
  pivot_longer(!gene_id, names_to = "sample_name", values_to = "counts_before") %>% mutate(merge = "before")

counts_merged_after <- counts(dds_all) %>% as.data.frame() %>% tibble::rownames_to_column(var = "gene_id") %>%
  pivot_longer(!gene_id, names_to = "sample_name", values_to = "counts_after") %>% mutate(merge = "after")

counts_joined <- full_join(counts_merged_after, counts_merged_before, by = c("gene_id","sample_name"))
```

```{r}
ggplot(filter(counts_joined, sample_name == "17mo-F-1")) + ggpointdensity::geom_pointdensity(aes(x = counts_before, y = counts_after)) + scale_x_log10() + scale_y_log10() + geom_abline(slope = 1, intercept = 0, color = "red", linetype = "dashed")
```
```{r}
count_diff <- counts_joined %>% mutate(diff = counts_before - counts_after, log2fc = log2(counts_before / counts_after)) 
range(count_diff$diff)
mn <- mean(count_diff$diff)
mdn <- median(count_diff$diff)

ggplot(count_diff,aes(x = log2fc)) + geom_histogram() + xlim(c(-5,5)) + ggtitle("difference in collapsing tech. reps. before vs after pseudoalignment")
ggplot(count_diff,aes(x = log2fc)) + geom_histogram(bins = 100) +
  xlim(c(-1,1)) +
  ggtitle("difference in collapsing tech. reps. before vs after pseudoalignment") + 
  annotate(geom = "text", label = glue::glue("mean log2fc: ", round(mn, 8 )), x = 0.5, y = 2e5) +
  annotate(geom = "text", label = glue::glue("median log2fc: ", mdn ), x = 0.5, y = 1.8e5)

pdf(here("plots/technical_replicate_collapsing_comparison.pdf"), width = 10, height = 10)
ggplot(count_diff,aes(x = log2fc)) + geom_histogram(bins = 100) +
  xlim(c(-1,1)) +
  ggtitle("difference in collapsing tech. reps. before vs after pseudoalignment") +
  facet_wrap(~sample_name, ncol = 4)

```

