---
title: "Identify groups of DE genes with similar dynamics"
output: html_document
---


```{r, include= F}
library(tidyverse)
library(glue)
library(DESeq2)
library(here)
library(tximport)
library(RColorBrewer)
library(vsn)
library(ComplexHeatmap)
library(biomaRt)
```

Read in experimental metadata
```{r}
data_dir <- here("kallisto_out_merged/")
metadata_path <- here("metadata_merged.csv")
list.files(data_dir)


exp_metadata <- read.csv(metadata_path, header = T)

# named_runs <- exp_metadata %>% deframe
```



Differential expression on relevant conditions
```{r}
dds <- readRDS(here("data/bulkrna_dds_collapsed_before.rds"))

test_names <- paste0("age_",c("6mo","17mo"),"_vs_P30")

res_list <- map(test_names, function(name){
  tbl <- read.csv(paste0(here("results/diff_exp_collapsed_before/"), name,".csv"), row.names = 1)
})

clip_nlog10p <- 10
sigthresh_nlog10p <- -log10(.05)
res_df <- map_df(test_names, function(name){
  tbl <- read.csv(paste0(here("results/diff_exp_collapsed_before/"), name,".csv"), row.names = 1)
  tbl <- tbl %>% mutate(condition = name,
                        nlog10p = ifelse(-log10(padj) > clip_nlog10p, clip_nlog10p, -log10(padj)),
                        nlog10p_clipped = ifelse(-log10(padj) > clip_nlog10p, T, F ))
  return(tbl)
})


```



Get biomart annotations for gene short names
```{r}
library(biomaRt)
#get gene names, make sure to use the right ensembl archive for genome used for alignment
#listMarts() 
ensembl <- useMart("ensembl")
datasets <- listDatasets(ensembl)
ensembl <- useDataset("mmusculus_gene_ensembl",mart=ensembl)

gene_id_trimmed <- rownames(dds)
gene_description <- getBM(attributes = c("ensembl_gene_id_version","external_gene_name","description","chromosome_name"),
                   filters = "ensembl_gene_id", values =gene_id_trimmed,  mart = ensembl)

geneid2name <- gene_description[,c("ensembl_gene_id_version", "external_gene_name")] %>% 
  mutate(ensembl_gene_id_trimmed = str_split_fixed(ensembl_gene_id_version, "\\.", 2) %>% .[,1])

gene_ids <- geneid2name %>% mutate(ensemblgene_id = ensembl_gene_id_version)
```
```{r}
pcut <- 0.05
#all genes that are DE in either condition
diff_genes <-res_df %>% filter(padj < pcut) %>% pull(ensemblgene_id) %>% unique()

res_df_filtered <- res_df %>% filter(ensemblgene_id %in% diff_genes)
res_df_filtered_wide <- res_df_filtered  %>%
  dplyr::select(c(ensemblgene_id, padj, nlog10p, condition, log2FoldChange)) %>%
  pivot_wider(names_from = condition, values_from = c(padj, log2FoldChange, nlog10p)) %>% 
  mutate(both_sig = padj_age_6mo_vs_P30 < pcut & padj_age_17mo_vs_P30<pcut,
         doe_17mo_vs_P30 = sign(log2FoldChange_age_17mo_vs_P30),
         doe_6mo_vs_P30 = sign(log2FoldChange_age_6mo_vs_P30),
         direction_consistent = (doe_17mo_vs_P30 == doe_6mo_vs_P30),
         direction = ifelse(direction_consistent, sign(log2FoldChange_age_6mo_vs_P30), paste(doe_6mo_vs_P30,doe_17mo_vs_P30, sep = ","))) 

#for genes significant for both timepoints, what are the direction of effects?
res_df_filtered_wide %>% filter(both_sig == T) %>% pull(direction) %>% table

#for genes significant for at least one timepoint, what are the direction of effects (over both timepoints)?
res_df_filtered_wide %>% pull(direction) %>% table

genes_down <- res_df_filtered_wide %>% filter(both_sig == T, direction == -1) %>% pull(ensemblgene_id) 
genes_up <- res_df_filtered_wide %>% filter(both_sig == T, direction == 1) %>% pull(ensemblgene_id) 

```

```{r}


ggplot(res_df_filtered_wide,aes(x = nlog10p_age_6mo_vs_P30, y=nlog10p_age_17mo_vs_P30 )) + geom_point(size = I(0.01))+ xlab("6mo vs P30 nlog10p") + ylab("17mo vs P30 nlog10p") + geom_hline(yintercept = sigthresh_nlog10p, linetype = "dashed", color = "red")+ geom_vline(xintercept = sigthresh_nlog10p, linetype = "dashed", color = "Red")

ggplot(res_df_filtered_wide,aes(x = log2FoldChange_age_6mo_vs_P30, y=log2FoldChange_age_17mo_vs_P30, color = interaction(both_sig,direction, sep = ":" ))) + geom_point(size = I(0.01))+ xlab("6mo vs P30 log2FC") + ylab("17mo vs P30 log2FC") + ggtitle("Subset of genes for which at least one condition is signficantly DE")+ guides(colour = guide_legend(override.aes = list(size=5)))
# 
# diff_genes_df <- res_df %>% group_by(condition) %>% mutate(sig = padj < pcut, lfc_direction = ifelse(log2FoldChange > 0, "up", "down")) %>% filter(ensemblgene_id %in% diff_genes) 

```
```{r}
genes_both_sig <- res_df_filtered_wide %>% filter(both_sig == T) %>% pull(ensemblgene_id)
#diff_genes has genes that are DE in at least one timepoint

#clustering on LFC
#on diff genes
log2fc_mat <- res_df_filtered_wide %>% dplyr::select(ensemblgene_id, log2FoldChange_age_6mo_vs_P30, log2FoldChange_age_17mo_vs_P30) %>% mutate(log2FoldChange_age_P30_vs_P30 = 0) %>% column_to_rownames(var = "ensemblgene_id") %>% as.matrix()

# clusters <- hclust(dist(log2fc_mat))
# plot(clusters)
```

```{r}
#Detect when genes have 0 variance for at least one timepoint, these LFC are estimates and magnitude dominates
counts_long <- assay(dds)[diff_genes,] %>% as.data.frame() %>% rownames_to_column(var = "ensemblgene_id") %>%
  pivot_longer(!ensemblgene_id, names_to = "sample_name", values_to = "count") %>% 
  separate(sample_name, c("age","sex","batch"), sep = "-", remove = F) 

var_df <- counts_long %>% group_by(age, ensemblgene_id) %>% summarise(gene_var = var(count)) %>% arrange(ensemblgene_id) 
var_df_wide <- var_df %>% pivot_wider(names_from = age, values_from = gene_var, names_glue = "gene_var.{age}")
zero_var_genes <- var_df %>% filter(gene_var == 0) %>% pull(ensemblgene_id) %>% unique() 

log2fc_mat_filtered <- log2fc_mat[!(rownames(log2fc_mat) %in% zero_var_genes),]
log2fc_mat_zerovar <- log2fc_mat[zero_var_genes,]

#Elbow Method for finding the optimal number of clusters
set.seed(123)
# Compute and plot wss for k = 2 to k = 15.
k.max <- 20
#all genes
wss <- sapply(1:k.max, 
              function(k){kmeans(log2fc_mat, k, nstart=50,iter.max = 15 )$tot.withinss})
plot(1:k.max, wss,
     type="b", pch = 19, frame = FALSE, 
     xlab="Number of clusters K",
     ylab="Total within-clusters sum of squares", main = "all genes")

#filtered genes
wss <- sapply(1:k.max, 
              function(k){kmeans(log2fc_mat_filtered, k, nstart=50,iter.max = 15 )$tot.withinss})
plot(1:k.max, wss,
     type="b", pch = 19, frame = FALSE, 
     xlab="Number of clusters K",
     ylab="Total within-clusters sum of squares", main = "filtered genes")

#zero_var genes
wss <- sapply(1:k.max, 
              function(k){kmeans(log2fc_mat_zerovar, k, nstart=50,iter.max = 15 )$tot.withinss})
plot(1:k.max, wss,
     type="b", pch = 19, frame = FALSE, 
     xlab="Number of clusters K",
     ylab="Total within-clusters sum of squares", main = "zero-var genes")
```

Clustering on all genes
```{r}
set.seed(114)
nclust <- 14
cluster.kmeans <- kmeans(log2fc_mat, centers = nclust)
cluster_df <- cluster.kmeans$cluster %>% enframe(name = "ensemblgene_id", value = "cluster.k.8") %>% add_count(cluster.k.8) %>% mutate(cluster_title = paste0("cluster ",cluster.k.8, " (n=",n,")"))

log2fc_long <- res_df_filtered %>%
  full_join(cluster_df) %>%
  mutate(condition = 
           case_when(condition == "age_6mo_vs_P30" ~ "6mo", condition == "age_17mo_vs_P30" ~ "17mo")) %>%
  mutate(transparency = 2*sqrt(1/n)) %>%
  mutate(transparency_corrected = pmin(1, transparency)) %>% 
  dplyr::select(-c(baseMean, lfcSE, stat, pvalue, padj, nlog10p, nlog10p_clipped))


p30_df <- log2fc_long %>% dplyr::select(external_gene_name, ensemblgene_id, cluster.k.8, n, cluster_title, transparency, transparency_corrected) %>% unique %>% mutate("log2FoldChange" = 0, condition = "P30")

log2fc_long <- rbind(log2fc_long, p30_df) %>% mutate(condition = factor(condition, levels = c("P30","6mo","17mo"))) 
dummy_df <- log2fc_long %>% dplyr::select(-cluster_title)

ggplot(log2fc_long, aes( x = condition, y = log2FoldChange, group = ensemblgene_id, alpha = transparency_corrected)) + geom_line(color = "black") + facet_wrap(~cluster_title, scales = "free_y") + ylab("log2FC estimates relative to P30") 

# ggplot(log2fc_long, aes( x = condition, y = log2FoldChange, group = ensemblgene_id)) + geom_line(data = dummy_df, color = "grey50")  + geom_line() + facet_wrap(~cluster_title) 

```

Is clustering (of extreme LFCs) driven by expression variance at a certain timepoint? (all genes)
```{r, eval = F}
counts_long <- assay(dds)[diff_genes,] %>% as.data.frame() %>% rownames_to_column(var = "ensemblgene_id") %>%
  pivot_longer(!ensemblgene_id, names_to = "sample_name", values_to = "count") %>% 
  separate(sample_name, c("age","sex","batch"), sep = "-", remove = F) 

var_df <- counts_long %>% group_by(age, ensemblgene_id) %>% summarise(gene_var = var(count)) %>% arrange(ensemblgene_id) 
var_df_wide <- var_df %>% pivot_wider(names_from = age, values_from = gene_var, names_glue = "gene_var.{age}")
zero_var_genes <- var_df %>% filter(gene_var == 0) %>% pull(ensemblgene_id) %>% unique() 

log2fc_long <- full_join(log2fc_long, var_df_wide) %>%
  mutate(zero_var = factor(ifelse(ensemblgene_id %in% zero_var_genes, TRUE, FALSE)))

ggplot(log2fc_long, aes( x = condition, y = log2FoldChange, group = ensemblgene_id, alpha = transparency_corrected, color = gene_var.P30)) + geom_line() + facet_wrap(~cluster_title, scales = "free_y") + ylab("log2FC estimates relative to P30") + scale_color_viridis_c(trans = "log10", na.value = "grey")

ggplot(log2fc_long, aes( x = condition, y = log2FoldChange, group = ensemblgene_id, alpha = transparency_corrected, color = gene_var.6mo)) + geom_line() + facet_wrap(~cluster_title, scales = "free_y") + ylab("log2FC estimates relative to P30") + scale_color_viridis_c(trans = "log10", na.value = "grey")

ggplot(log2fc_long, aes( x = condition, y = log2FoldChange, group = ensemblgene_id, alpha = transparency_corrected, color = gene_var.17mo)) + geom_line() + facet_wrap(~cluster_title, scales = "free_y") + ylab("log2FC estimates relative to P30") + scale_color_viridis_c(trans = "log10", na.value = "grey")

ggplot(log2fc_long, aes( x = condition, y = log2FoldChange, group = ensemblgene_id, alpha = transparency_corrected, color = zero_var)) + geom_line() + facet_wrap(~cluster_title, scales = "free_y") + ylab("log2FC estimates relative to P30") 

set.seed(114)
log2fc_mat_gene_var <- log2fc_mat %>% as.data.frame() %>%
  rownames_to_column(var = "ensemblgene_id") %>% 
  mutate(zero_var = factor(ifelse(ensemblgene_id %in% zero_var_genes, TRUE, FALSE)))

nclust <- 14
cluster.kmeans <- kmeans(log2fc_mat, centers = nclust)
cluster_df <- cluster.kmeans$cluster %>% enframe(name = "ensemblgene_id", value = "cluster.k.14.var") %>% add_count(cluster.k.14.var, name = "n.var") %>% mutate(cluster_title.var = paste0("cluster ",cluster.k.14.var, " (n=",n.var,")"))

log2fc_long <- log2fc_long %>%
  left_join(cluster_df)

ggplot(log2fc_long, aes( x = condition, y = log2FoldChange, group = ensemblgene_id, alpha = transparency_corrected, color = zero_var)) + geom_line() + facet_wrap(~cluster_title, scales = "free_y") + ylab("log2FC estimates relative to P30") + ggtitle("clustering on LFC")

ggplot(log2fc_long, aes( x = condition, y = log2FoldChange, group = ensemblgene_id, alpha = transparency_corrected, color = zero_var)) + geom_line() + facet_wrap(~cluster_title.var, scales = "free_y") + ylab("log2FC estimates relative to P30") + ggtitle("clustering on LFC and binary zero-var indicator")

```

Clustering on zero-var 
TODO: What drives the magnitude of the LFC estimates with a denom/numer of 0? is it expression magnitude or variance?
-if these are not important, I can scale the matrix before clustering to get dynamics separately from magnitude
```{r}
set.seed(114)
nclust <- 7
#either cluster on log2fc matrix or scaled log2fc matrix
cluster.kmeans <- kmeans(t(scale(t(log2fc_mat_zerovar), center = F, scale = T)), centers = nclust)
cluster_df <- cluster.kmeans$cluster %>% enframe(name = "ensemblgene_id", value = "cluster.k.8") %>% add_count(cluster.k.8) %>% mutate(cluster_title = paste0("cluster ",cluster.k.8, " (n=",n,")"))

log2fc_long <- res_df_filtered %>%
  filter(ensemblgene_id %in% zero_var_genes) %>%
  full_join(cluster_df) %>%
  mutate(condition = 
           case_when(condition == "age_6mo_vs_P30" ~ "6mo", condition == "age_17mo_vs_P30" ~ "17mo")) %>%
  mutate(transparency = sqrt(1/n)) %>%
  mutate(transparency_corrected = pmin(1, transparency)) %>% 
  dplyr::select(-c(baseMean, lfcSE, stat, pvalue, padj, nlog10p, nlog10p_clipped))


p30_df <- log2fc_long %>% dplyr::select(external_gene_name, ensemblgene_id, cluster.k.8, n, cluster_title, transparency, transparency_corrected) %>% unique %>% mutate("log2FoldChange" = 0, condition = "P30")

log2fc_long <- rbind(log2fc_long, p30_df) %>% mutate(condition = factor(condition, levels = c("P30","6mo","17mo"))) 
dummy_df <- log2fc_long %>% dplyr::select(-cluster_title)

ggplot(log2fc_long, aes( x = condition, y = log2FoldChange, group = ensemblgene_id, alpha = transparency_corrected)) + geom_line(color = "black") + facet_wrap(~cluster_title, scales = "free_y") + ylab("log2FC estimates relative to P30") 

# ggplot(log2fc_long, aes( x = condition, y = log2FoldChange, group = ensemblgene_id)) + geom_line(data = dummy_df, color = "grey50")  + geom_line() + facet_wrap(~cluster_title) 

```

Clustering on non-zero var genes
```{r}
set.seed(114)
nclust <- 6
cluster.kmeans <- kmeans(t(scale(t(log2fc_mat_filtered), scale = T, center = F)), centers = nclust)
cluster_df <- cluster.kmeans$cluster %>% enframe(name = "ensemblgene_id", value = "cluster.k.8") %>% add_count(cluster.k.8) %>% mutate(cluster_title = paste0("cluster ",cluster.k.8, " (n=",n,")"))

log2fc_long <- res_df_filtered %>%
  filter(!(ensemblgene_id %in% zero_var_genes)) %>%
  full_join(cluster_df) %>%
  mutate(condition = 
           case_when(condition == "age_6mo_vs_P30" ~ "6mo", condition == "age_17mo_vs_P30" ~ "17mo")) %>%
  mutate(transparency = 2*sqrt(1/n)) %>%
  mutate(transparency_corrected = pmin(1, transparency)) %>% 
  dplyr::select(-c(baseMean, lfcSE, stat, pvalue, padj, nlog10p, nlog10p_clipped))


p30_df <- log2fc_long %>% dplyr::select(external_gene_name, ensemblgene_id, cluster.k.8, n, cluster_title, transparency, transparency_corrected) %>% unique %>% mutate("log2FoldChange" = 0, condition = "P30")

log2fc_long <- rbind(log2fc_long, p30_df) %>% mutate(condition = factor(condition, levels = c("P30","6mo","17mo"))) 
dummy_df <- log2fc_long %>% dplyr::select(-cluster_title)

ggplot(log2fc_long, aes( x = condition, y = log2FoldChange, group = ensemblgene_id, alpha = transparency_corrected)) + geom_line(color = "black") + facet_wrap(~cluster_title, scales = "free_y") + ylab("log2FC estimates relative to P30") 

# ggplot(log2fc_long, aes( x = condition, y = log2FoldChange, group = ensemblgene_id)) + geom_line(data = dummy_df, color = "grey50")  + geom_line() + facet_wrap(~cluster_title) 

```


```{r}
#clustering on shrinkage LFC
resLFC6mo <- lfcShrink(dds,coef = "age6mo", type="apeglm") %>%
  as.data.frame() %>% .[, "log2FoldChange", drop = F] %>%
  rownames_to_column(var = "ensemblgene_id")
resLFC17mo <- lfcShrink(dds,coef = "age17mo", type="apeglm") %>%
  as.data.frame() %>% .[, "log2FoldChange", drop = F] %>%
  rownames_to_column(var = "ensemblgene_id")

resLFC <- full_join(resLFC6mo,resLFC17mo, suffix = c(".6mo",".17mo"), by = "ensemblgene_id")
ggplot(resLFC, aes(x = log2FoldChange.6mo, y= log2FoldChange.17mo)) + geom_point()

#on diff genes
log2fc_mat <- resLFC %>% filter(ensemblgene_id %in% diff_genes) %>%
  mutate(log2FoldChange.P30 = 0) %>% column_to_rownames(var = "ensemblgene_id") %>% as.matrix()

# clusters <- hclust(dist(log2fc_mat))
# plot(clusters)

nclust <- 8
cluster.kmeans <- kmeans(log2fc_mat, centers = nclust)
cluster_df <- cluster.kmeans$cluster %>% enframe(name = "ensemblgene_id", value = "cluster.k.8") %>% add_count(cluster.k.8) %>% mutate(cluster_title = paste0("cluster ",cluster.k.8, " (n=",n,")"))

log2fc_long <- log2fc_mat %>% as.data.frame() %>%
  rownames_to_column(var = "ensemblgene_id") %>%
  full_join(cluster_df) %>%
  pivot_longer(!c(ensemblgene_id, n, cluster_title, cluster.k.8), values_to = "logFCshrink", names_to = "condition") %>%
  mutate(condition = case_when(
    condition == "log2FoldChange.6mo" ~ "6mo",
    condition == "log2FoldChange.17mo" ~ "17mo", 
    condition == "log2FoldChange.P30" ~ "P30"
  ))%>%
  mutate(condition = factor(condition, levels = c("P30","6mo","17mo")))

  
dummy_df <- log2fc_long %>% dplyr::select(-cluster_title)
ggplot(log2fc_long, aes( x = condition, y = logFCshrink, group = ensemblgene_id)) + geom_line(color = "black") + facet_wrap(~cluster_title, scales = "free_y") + ylab("log2FC shrinkage estimates relative to P30") 

ggplot(log2fc_long, aes( x = condition, y = logFCshrink, group = ensemblgene_id)) + geom_line(data = dummy_df, color = "grey50")  + geom_line() + facet_wrap(~cluster_title) 
```

```{r, eval = F}
res <- results(dds, name = "age17mo")
res %>% head
summary(res)

resultsNames(dds)
resLFC <- lfcShrink(dds,coef = "age17mo", type="apeglm")

assertthat::assert_that(all(rownames(res) == rownames(resLFC)))
qplot(res6mo$log2FoldChange, resLFC17mo$log2FoldChange, color = res$baseMean) + scale_color_viridis_c(name = "res$baseMean", trans = "log")
qplot(res$log2FoldChange, resLFC$log2FoldChange.6mo, color = res$lfcSE) + scale_color_viridis_c(name = "res$lfcSE", trans = "log")

```







