---
title: "Projections"
author: "Jared Slosberg"
date: "11/21/2022"
output: html_document
---

```{r, include= F}
library(tidyverse)
library(glue)
library(DESeq2)
library(here)
library(tximport)
library(RColorBrewer)
library(vsn)
library(ComplexHeatmap)
library(biomaRt)
```

Read in experimental metadata
```{r}
metadata_path <- here("metadata_merged_filtered.csv")


exp_metadata <- read.csv(metadata_path, header = T)

#subset to intestine samples (remove fat and stomach)\

# named_runs <- exp_metadata %>% deframe
```

Read in DESEQ2 object
```{r}
dds <- readRDS(here("data/bulkrna_dds.rds"))
```

Read in learned and other expression signatures
```{r}
lmmp_gw <- read.csv("/data/users/jared/ENS/6mo_LMMP/results/NMF/lmmp/old_pattern_run/50dims/pattern_gene_weights.csv", row.names = 1 )

lmmp_gw_clean <- lmmp_gw 

head(lmmp_gw_clean)
```

Get biomart annotations for gene short names
```{r}
library(biomaRt)
#get gene names, make sure to use the right ensembl archive for genome used for alignment
#listMarts() 
ensembl <- useMart("ensembl")
datasets <- listDatasets(ensembl)
ensembl <- useDataset("mmusculus_gene_ensembl",mart=ensembl)

gene_id_trimmed <- stringr::str_split_fixed(rownames(dds),"\\.",2)[,1]
gene_description <- getBM(attributes = c("ensembl_gene_id_version","external_gene_name","description","chromosome_name"),
                   filters = "ensembl_gene_id", values =gene_id_trimmed,  mart = ensembl)

geneid2name <- gene_description[,c("ensembl_gene_id_version", "external_gene_name")] %>% 
  mutate(ensembl_gene_id_trimmed = str_split_fixed(ensembl_gene_id_version, "\\.", 2) %>% .[,1])

```

```{r}
library(projectR)

vsd <- vst(dds, blind=FALSE)
vsd_cleaned <- as.matrix(assay(vsd)) %>% as.data.frame() %>% rownames_to_column(var = "ensembl_gene_id_trimmed") %>% left_join(., geneid2name) %>% drop_na() %>% dplyr::select(-starts_with("ensembl")) %>% filter(!duplicated(external_gene_name)) %>% column_to_rownames(var = "external_gene_name") %>% as.matrix()

set.seed(007)
proj_res <- projectR(vsd_cleaned, as.matrix(lmmp_gw_clean))

proj_long <- proj_res %>% as.data.frame() %>% rownames_to_column(var = "pattern") %>% pivot_longer(!pattern,names_to = "sample_name", values_to = "weight") %>%
  left_join(., exp_metadata)

proj_long <- proj_long %>% mutate(
  age = factor(age, levels = c("P30","6mo","17mo")),
  age_numeric = case_when(age == "17mo" ~ 510,
                          age == "6mo" ~ 180 ,
                          age == "P30" ~ 30))
```

```{r}
pattern_names <- colnames(lmmp_gw_clean)
lm_res <- map_df(pattern_names, function(patt){
  fit <- lm(weight ~ age_numeric, filter(proj_long, pattern == patt))
  data.frame(pattern = patt, coef = coefficients(fit)["age_numeric"],
             pvalue = summary(fit)$coefficients["age_numeric",4])
}) %>% mutate(sig = pvalue < 0.05)

ggplot(lm_res, aes(x = pattern, y = coef)) + geom_bar(stat = "identity") + 
  theme(axis.text.x = element_text(angle =90)) +
  geom_text(data = filter(lm_res, sig == T), aes(x = pattern, y = coef),label = "*", color = "red") +
  ggtitle("Coefficient for age in pattern_weight ~ numeric(age): 6mo LMMP patterns")

#with sex included - removed term "sexM:age17mo" since sample size = 2
terms <- c("age_numeric", "sexM" , "age_numeric:sexM")
lm_res <- map_df(pattern_names, function(patt){
  
  fit <- lm(weight ~ age_numeric + sex + age_numeric:sex, filter(proj_long, pattern == patt))
  
  map_df(terms, function(term){
    test <- data.frame(pattern = patt, model_term = term, coef = coefficients(fit)[term],
             pvalue = summary(fit)$coefficients[term,4])
  })
}) %>% mutate(significance = case_when(
  sign(coef) == 1 & pvalue < 0.05 ~ "up",
  sign(coef) == -1 & pvalue < 0.05 ~ "down",
  .default = NA
  ))

coef_sig_plot_list <- map(terms, function(term){
  res <- lm_res %>% filter(model_term == term )
  
  ggplot(res, aes(x = pattern, y = coef, fill =significance)) + geom_bar(stat = "identity") + 
  theme(axis.text.x = element_text(angle =90)) +
 # geom_text(data = filter(res, sig == T), aes(x = pattern, y = coef),label = "*", color = "red") +
  ggtitle(paste0("Coefficient for ", term, " in pattern_weight ~ numeric(age) + sex + numeric(age):sex - 6mo LMMP patterns")) + 
    scale_fill_manual(values = c("up" = "red", "down" = "blue"))
          
})

coef_sig_plot_list

```

```{r}

pattern_var <- rowVars(proj_res) %>% setNames(pattern_names) %>% enframe(name = "pattern", value = "var")
pattern_mean <- rowMeans(proj_res) %>% setNames(pattern_names) %>% enframe(name = "pattern", value = "mean")

pattern_df <- full_join(pattern_mean, pattern_var) %>% mutate(var_pct = var/abs(mean))


ggplot(pattern_df, aes(x = mean, y = var)) + geom_point()+ ggrepel::geom_text_repel(aes(label = pattern))
ggplot(pattern_df, aes(x = mean, y = var_pct)) + geom_point() + ggrepel::geom_text_repel(aes(label = pattern))

```


```{r}
#pl <- ggplot(proj_long) + geom_point(aes(x = age, y = weight, color = sample_name)) + facet_wrap(~pattern, ncol = 5, scales = "free_y")


pl <- ggplot(proj_long,aes(x = age_numeric, y = weight, group = sex, color = sex))  +
  stat_summary(fun = mean, geom = "line",position=position_dodge2(width = .3)) + 
  stat_summary(fun.data = mean_se, geom = "errorbar",position=position_dodge2(width = 0.2), width = .3) +
  geom_point(size = 0.2, position = position_dodge(width=.3)) +
  facet_wrap(~pattern, ncol = 5, scales = "free_y") + 
  theme(axis.text.y = element_blank()) +


# pl <- ggplot(proj_long,aes(x = as.factor(age_numeric), y = weight, as.factor(age_numeric), color = sex))  +
#   # stat_summary(fun = mean, geom = "point",position=position_dodge2(width = .3)) + 
#   # stat_summary(fun.data = mean_se, geom = "errorbar",position=position_dodge2(width = 0.2), width = .3) +
#   #geom_point(size = 0.2, position = position_dodge(width=.3)) +
#   geom_boxplot() +
#   facet_wrap(~pattern, ncol = 5, scales = "free_y") + 
#   theme(axis.text.y = element_blank()) +
#   geom_smooth(method= "lm")

pl
```

Changes to pipeline

```{r}
total_gene_sums <- vsd_cleaned %>% rowSums()
hist(total_gene_sums)

hist(as.numeric(as.matrix(lmmp_gw_clean)))
ggplot(data.frame(x = unlist(lmmp_gw_clean))) + geom_histogram(aes(x)) + scale_y_log10()

rowData(dds)[,"ensembl_gene_id_trimmed"] <- rownames(rowData(dds))
gene_df <- rowData(dds)[,c("external_gene_name","ensembl_gene_id_trimmed")] %>% as.data.frame()

mt_genes_ind <- str_detect(gene_df$external_gene_name, "^mt-") %>% which
mt_genes <- gene_df[mt_genes_ind,]
ribo_genes_ind <- str_detect(gene_df$external_gene_name, "^Rps|^Rpl") %>% which()
ribo_genes <- gene_df[ribo_genes_ind,]
hemo_genes_ind <- str_detect(gene_df$external_gene_name, "^Hbb|^Hba") %>% which()
hemo_genes <- gene_df[hemo_genes_ind,]

signal <- sum(assay(dds, "counts"))
mt_signal <- sum(assay(dds, "counts")[mt_genes_ind,])
mt_ratio <- 100*(mt_signal/signal)
ribo_signal <- sum(assay(dds, "counts")[ribo_genes_ind,])
ribo_ratio <- 100*(ribo_signal/signal)

hemo_signal <- sum(assay(dds, "counts")[hemo_genes_ind,])
hemo_ratio <- 100*(hemo_signal/signal)

remove_genes_ind <- c(ribo_genes_ind, mt_genes_ind)
remove_genes_ensembl <- c(ribo_genes$ensembl_gene_id_trimmed, mt_genes$ensembl_gene_id_trimmed)
keep_genes_ensembl <- gene_df[!(gene_df$ensembl_gene_id_trimmed %in% remove_genes_ensembl),"ensembl_gene_id_trimmed"]

data.frame(mt_ratio, ribo_ratio, hemo_ratio)

```

Together, mitochondrial and ribosomal genes make up ~27% of the counts in the data. Remove this to keep high variance genes for projection

```{r}
set.seed(007)
vsd_filtered <- as.matrix(assay(vsd)[keep_genes_ensembl,]) %>% as.data.frame() %>% rownames_to_column(var = "ensembl_gene_id_trimmed") %>% left_join(., geneid2name) %>% drop_na() %>% dplyr::select(-starts_with("ensembl")) %>% filter(!duplicated(external_gene_name)) %>% column_to_rownames(var = "external_gene_name") %>% as.matrix()

proj_res_filtered <- projectR(vsd_filtered, as.matrix(lmmp_gw_clean))

#write.csv(proj_res, here("results/NMF/lmmp_projection/bulk_in_tc_lmmp.csv"), quote = F, row.names = T)

proj_long <- proj_res %>% as.data.frame() %>% rownames_to_column(var = "pattern") %>% pivot_longer(!pattern,names_to = "sample_name", values_to = "weight") %>%
  left_join(., exp_metadata)

proj_long <- proj_long %>% mutate(
  age = factor(age, levels = c("P30","6mo","17mo")),
  age_numeric = case_when(age == "17mo" ~ 510,
                          age == "6mo" ~ 180 ,
                          age == "P30" ~ 30))

pattern_var <- rowVars(proj_res_filtered) %>% setNames(pattern_names) %>% enframe(name = "pattern", value = "var")
pattern_mean <- rowMeans(proj_res_filtered) %>% setNames(pattern_names) %>% enframe(name = "pattern", value = "mean")

pattern_df <- full_join(pattern_mean, pattern_var) %>% mutate(var_pct = var/abs(mean))


ggplot(pattern_df, aes(x = mean, y = var)) + geom_point()+ ggrepel::geom_text_repel(aes(label = pattern))
ggplot(pattern_df, aes(x = mean, y = var_pct)) + geom_point() + ggrepel::geom_text_repel(aes(label = pattern))
```

Didn't alleviate the problem. Pattern 11 is also a RBC pattern. Remove patterns 7,11,44 before projection

```{r}
set.seed(007)

gw_filt <- lmmp_gw_clean %>% dplyr::select(-paste0("cellPattern", c(7,11,44)))
patt_names_filt <- colnames(gw_filt)


proj_res_filtered <- projectR(vsd_filtered, as.matrix(gw_filt))

write.csv(proj_res_filtered, here("results/NMF/lmmp_projection/bulk_in_tc_lmmp.csv"), quote = F, row.names = T)

proj_long_filtered <- proj_res_filtered %>% as.data.frame() %>% rownames_to_column(var = "pattern") %>% pivot_longer(!pattern,names_to = "sample_name", values_to = "weight") %>%
  left_join(., exp_metadata)

proj_long_filtered <- proj_long_filtered %>% mutate(
  age = factor(age, levels = c("P30","6mo","17mo")),
  age_numeric = case_when(age == "17mo" ~ 510,
                          age == "6mo" ~ 180 ,
                          age == "P30" ~ 30))

pattern_var <- rowVars(proj_res_filtered) %>% setNames(patt_names_filt) %>% enframe(name = "pattern", value = "var")
pattern_mean <- rowMeans(proj_res_filtered) %>% setNames(patt_names_filt) %>% enframe(name = "pattern", value = "mean")

pattern_df <- full_join(pattern_mean, pattern_var) %>% mutate(var_pct = var/abs(mean))


ggplot(pattern_df, aes(x = mean, y = var)) + geom_point()+ ggrepel::geom_text_repel(aes(label = pattern))
ggplot(pattern_df, aes(x = mean, y = var_pct)) + geom_point() + ggrepel::geom_text_repel(aes(label = pattern))

#pca
projection_pc <- prcomp(t(proj_res_filtered), scale = T)
sample_weights <- projection_pc$x %>% as.data.frame() %>% tibble::rownames_to_column(var = "sample_name") %>% left_join(exp_metadata)

sample_umap <- uwot::umap(as.data.frame( projection_pc$x),
                          n_neighbors = 2) %>% as.data.frame()
rownames(sample_umap) <- rownames(projection_pc$x)
colnames(sample_umap) <- c("UMAP1","UMAP2")
sample_umap <- sample_umap %>% tibble::rownames_to_column(var = "sample_name") %>% left_join(., exp_metadata) 

ggplot(sample_weights, shape = 21, aes(x = PC1, y = PC2, color = age, fill = sex)) + geom_point()
ggplot(sample_umap, shape = 21, aes(x = UMAP1, y = UMAP2, color = age, fill = sex)) + geom_point()
```

```{r}
pl <- ggplot(proj_long_filtered) + geom_point(aes(x = age, y = weight, color = sample_name)) + facet_wrap(~pattern, ncol = 5, scales = "free_y")


pl <- ggplot(proj_long_filtered,aes(x = age_numeric, y = weight, group = sex, color = sex))  +
  stat_summary(fun = mean, geom = "line",position=position_dodge2(width = .3)) + 
  stat_summary(fun.data = mean_se, geom = "errorbar",position=position_dodge2(width = 0.2), width = .3) +
  geom_point(size = 0.2, position = position_dodge(width=.3)) +
  facet_wrap(~pattern, ncol = 5, scales = "free_y") 
  #theme(axis.text.y = element_blank()) +

if(!dir.exists(here("plots/NMF/lmmp_projections"))){
  dir.create(here("plots/NMF/lmmp_projections"))
}

pdf(here("plots/NMF/lmmp_projections/projection_weights_by_sample.pdf"), height = 20, width = 15)
pl
dev.off()
```

```{r}
pattern_names <- colnames(gw_filt)
lm_res <- map_df(pattern_names, function(patt){
  fit <- lm(weight ~ age_numeric, filter(proj_long_filtered, pattern == patt))
  data.frame(pattern = patt, coef = coefficients(fit)["age_numeric"],
             pvalue = summary(fit)$coefficients["age_numeric",4])
}) %>% mutate(padj = min(1, pvalue*length(pattern_names)), 
              significance = case_when(
  sign(coef) == 1 & padj < 0.05 ~ "up",
  sign(coef) == -1 & padj < 0.05 ~ "down",
  .default = NA
  ))

ggplot(lm_res, aes(x = pattern, y = coef, fill =significance)) + geom_bar(stat = "identity") + 
  theme(axis.text.x = element_text(angle =90)) +
    scale_fill_manual(values = c("up" = "red", "down" = "blue")) +
  ggtitle("Coefficient for age in pattern_weight ~ numeric(age): 6mo LMMP patterns")

#with sex included - removed term "sexM:age17mo" since sample size = 2
terms <- c("age_numeric", "sexM" , "age_numeric:sexM")
lm_res <- map_df(pattern_names, function(patt){
  
  fit <- lm(weight ~ age_numeric + sex + age_numeric:sex, filter(proj_long, pattern == patt))
  
  map_df(terms, function(term){
    test <- data.frame(pattern = patt, model_term = term, coef = coefficients(fit)[term],
             pvalue = summary(fit)$coefficients[term,4])
  })
}) %>% mutate(padj = pmin(1, pvalue*length(pattern_names)), 
              significance = case_when(
  sign(coef) == 1 & padj < 0.05 ~ "up",
  sign(coef) == -1 & padj < 0.05 ~ "down",
  .default = NA
  ))

coef_sig_plot_list <- map(terms, function(term){
  res <- lm_res %>% filter(model_term == term )
  
  ggplot(res, aes(x = pattern, y = coef, fill =significance)) + geom_bar(stat = "identity") + 
  theme(axis.text.x = element_text(angle =90)) +
 # geom_text(data = filter(res, sig == T), aes(x = pattern, y = coef),label = "*", color = "red") +
  ggtitle(paste0("Coefficient for ", term, " in pattern_weight ~ numeric(age) + sex + numeric(age):sex - 6mo LMMP patterns")) + 
    scale_fill_manual(values = c("up" = "red", "down" = "blue"))
          
})

pdf(here("plots/NMF/lmmp_projections/projection_sig_by_term.pdf"), width = 10, height = 6)
coef_sig_plot_list
dev.off()

write.csv(lm_res, here("results/NMF/lmmp_projection/age_pattern_regression_lm.csv"), quote = F, row.names = F)
```
Even after removing a few dominating "RBC" patterns, nothing is significantly differentially used after multiple testing correction

```{r}
#nominally significant
lm_res %>% filter(pvalue < 0.05) %>% arrange(desc(coef))

m_samples <- exp_metadata %>% filter(sex == "M") %>% pull(sample_name)
f_samples <- exp_metadata %>% filter(sex == "F") %>% pull(sample_name)

m_exp <- vsd_filtered %>% as.data.frame() %>% dplyr::select(all_of(m_samples))
f_exp <- vsd_filtered %>% as.data.frame() %>% dplyr::select(all_of(f_samples))


library(projectR)
test<- projectionDriveR(m_exp, f_exp, gw_filt, pattern_name = "cellPattern41", normalize_pattern = F)
```

Projection of bulk signatures with lmmp scRNA. "Where are these genes expressed"

```{r, eval = F}
lmmp <- readRDS("/data/users/jared/ENS/Timecourse_ENS/TC_LMMP.rds")

#compare distributions of normalized counts to make sure projection is fair
as.matrix(assay(lmmp[,1:1000], "logcounts")) %>% hist(breaks = 100)
as.matrix(assay(lmmp[,1:1000], "logcounts")) %>% unlist() %>% .[. > 1] %>% hist(breaks = 100)
as.matrix(assay(vsd)) %>% hist(breaks = 100)
```

```{r, eval = F}

dimensionalities <- c("10","15","20","25")
nmf_list <- map(dimensionalities, function(dimensionality){
  
    gene_weights <- read.csv(here("results/NMF", paste0("dim",dimensionality), "gene_weights.csv"),row.names = 1)
    sample_weights <- read.csv(here("results/NMF", paste0("dim",dimensionality), "sample_weights.csv"), row.names = 1)
  
  return(list(A = sample_weights, P = gene_weights))
}) %>% setNames(dimensionalities)
```

```{r, eval = F}
source("/data/users/jared/ENS/Timecourse_ENS/scripts/accessory_functions/pattern_plotting.R")

umap_coord <- reducedDim(lmmp, "UMAP") %>% as.data.frame() %>% rownames_to_column(var = "cell_id") 
colnames(umap_coord) <- c("cell_id", "UMAP1","UMAP2")

nmf_list <- map(dimensionalities, function(dimensionality){
  
    gw <- nmf_list[[dimensionality]][["P"]] %>% as.matrix()
    proj_res <- projectR(as.matrix(assay(lmmp, "logcounts")), gw) 
    
    # proj_long <- proj_res %>%
    #   as.data.frame() %>% 
    #   rownames_to_column(var = "pattern") %>%
    #   pivot_longer(!pattern,names_to = "cell_id", values_to = "weight") %>%
    #   left_join(., umap_coord)
    
    pData(lmmp) <- pData(lmmp) %>% as.data.frame() %>% rownames_to_column(var = "cell_id") %>%
      left_join(., rownames_to_column(as.data.frame(t(proj_res)), "cell_id")) %>% column_to_rownames("cell_id") %>% DataFrame()
    
    embs <- lapply(1:as.numeric(dimensionality), plotCellPatterns, lmmp, do.clip = c(0,.98), pattern_prefix = "pattern")
    embs_rast <- map(embs, function(x){
      x %>% ggrastr::rasterise(dpi = 200)
      })
  
    
    pdf(here(paste0("plots/NMF/lmmp_projections/tc_lmmp_in_zasloff_patterns_k",dimensionality,".pdf")),
        width = 25, height = as.numeric(dimensionality)/2)

        print(ggpubr::ggarrange(plotlist = embs_rast, ncol = 5))
    dev.off()
    
    write.csv(proj_res, here(paste0("results/NMF/lmmp_projection/tc_lmmp_in_zasloff_patterns_k",dimensionality,".csv")), quote = F, row.names = T)

    return(proj_res)

}) %>% setNames(dimensionalities)


```

```{r, eval = F}
library(preprocessCore)


dat <- runif(1000)
hist(dat)
test <- qnorm((rank(dat,na.last="keep")-0.5)/sum(!is.na(dat)))
hist(test)

proj_res[1,1:1000] %>% hist
proj_res[3,1:1000] %>% qnorm((rank(.,na.last="FALSE")-0.5)/sum(!is.na(.))) %>% hist()
proj_res[3,1:1000] %>% qnorm((rank(.,na.last="FALSE")-0.5)/sum(!is.na(.))) %>% mean(na.rm = T)

x <- (rank(proj_res[1,1:1000] ,na.last="keep")-0.5)/sum(!is.na(proj_res[1,1:1000]))

```

