---
title: "Untitled"
author: "Jared Slosberg"
date: "4/28/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(DESeq2)
library(here)
library(tidyverse)
```

```{r}
db_path <- here("data/ligand_receptor_db/")
type <- c("ecm","protein_types","receptors","ligands")

signalling_list <- map(type, function(tp){
  read.delim(paste0(db_path, tp, ".txt")) %>% setNames(c("hgnc_symbol","type"))
}) %>% setNames(type)

all_human_genes <- signalling_list %>% bind_rows() %>% pull(hgnc_symbol) %>% unique
```


```{r}
library(biomaRt)
#get gene names, make sure to use the right ensembl archive for genome used for alignment
#listMarts() 
human = useMart("ensembl", dataset = "hsapiens_gene_ensembl", host = "https://dec2021.archive.ensembl.org/")
mouse = useMart("ensembl", dataset = "mmusculus_gene_ensembl", host = "https://dec2021.archive.ensembl.org/")


genesV2 = getLDS(attributes = c("hgnc_symbol"), filters = "hgnc_symbol", values = all_human_genes , mart = human, attributesL = c("mgi_symbol", "ensembl_gene_id"), martL = mouse, uniqueRows=T)
humanx <- unique(genesV2[, 2])

gene_description <- getBM(attributes = c("ensembl_gene_id_version","external_gene_name","description","chromosome_name"),
                   filters = "ensembl_gene_id", values =gene_id_trimmed,  mart = ensembl)

geneid2name <- gene_description[,c("ensembl_gene_id_version", "external_gene_name")] %>% 
  mutate(ensembl_gene_id_trimmed = str_split_fixed(ensembl_gene_id_version, "\\.", 2) %>% .[,1])

gene_ids <- geneid2name %>% mutate(ensemblgene_id = ensembl_gene_id_version)
```

